sqlglot/dialects/duckdb.py

```diff
@@ -738,3 +738,31 @@ def withingroup_sql(self, expression: exp.WithinGroup) -> str:
             this = self.sql(expression, "this").rstrip(")")
 
             return f"{this}{expression_sql})"
+
+        def length_sql(self, expression: exp.Length) -> str:
+            arg = expression.this
+
+            if not arg.type:
+                from sqlglot.optimizer.annotate_types import annotate_types
+
+                arg = annotate_types(arg)
+
+            if arg.is_type(exp.DataType.Type.VARCHAR) or arg.is_string:
+                return self.func("LENGTH", arg)
+
+            # Dialects like BQ and Snowflake also accept binary values, so if we can't determine the
+            # arg type then attempt case-when resolution
+            trycast_blob = exp.TryCast(this=arg, to=exp.DataType.build(exp.DataType.Type.VARBINARY))
+            trycast_varchar = exp.TryCast(
+                this=arg, to=exp.DataType.build(exp.DataType.Type.VARCHAR)
+            )
+
+            case = (
+                exp.case(self.func("TYPEOF", arg))
+                .when(
+                    "'VARCHAR'", exp.Anonymous(this="LENGTH", expressions=[trycast_varchar])
+                )  # anonymous to break length_sql recursion
+                .when("'BLOB'", self.func("OCTET_LENGTH", trycast_blob))
+            )
+
+            return self.sql(case)
```

Comment:
This will make `LENGTH(x)` generate as `CASE TYPEOF(x) WHEN 'VARCHAR' THEN LENGTH(TRY_CAST(x AS TEXT)) WHEN 'BLOB' THEN OCTET_LENGTH(TRY_CAST(x AS BLOB)) END` even for duckdb -> duckdb.

I'd rather we have an arg in `Length` that signifies whether binary values can also be used as arguments, and only do something else in duckdb if that arg is set properly.

Reply:
Good idea

Reply:
Make sure to handle that arg gracefully if it's a boolean, because otherwise we'll crash in `Generator.sql`. We basically need to find all places where `Length` is generated (+ overrides) and exclude the flag.